### ISSUE TITLE
nvmf target stops to listen when subsystem is destroyed

### ISSUE THREAD

**Issue by jkryl** [spdk/spdk#1129]
This happens when running the latest bits from v19.10.x branch.
There is a spdk_nvmf_tgt_listen() which opens a port for specified
transport (trid) which opens possibility to accept new connections
from initiators. However there is no counterpart of this function
(i.e. spdk_nvmf_tgt_stop_listen()), which would stop listening.
Instead the current code relies on spdk_nvmf_subsystem_destroy()
to stop the listener. That's IMHO a faulty design. Subsystems come
and go. And even if the last subsystem is removed, the target should
continue to listen. A target without subsystems is just as valid.
tgt should stop to listen only if explicitly told to do so or if it is destroyed.

An example of my use case when this becomes a problem is as follows. We have
nvmf target. spdk_nvmf_tgt_listen() is called when initializing the target.
Subsystems are dynamically created and destroyed as needed. TCP is used
as transport. When a subsystem is destroyed, the target becomes inaccessible
(port is closed). Even if I create a new subsystem afterwards, the target is still
inaccessible. I would have to call spdk_nvmf_tgt_listen() again which does not
make sense to me.

I have put together a quick fix which we use for our project. I can make a long term
fix for spdk out of it, if we agree that this is the best way how to fix it.

**Comment by benlwalker** [comment]
Agreed. There probably needs to be an RPC interface to explicitly create and destroy a listener as well. Right now the RPCs do it automatically when listeners are added to subsystems, but that should be a simpler access control operation instead (just like setting the allowed hosts).

**Comment by jkryl** [comment]
@benlwalker : Thanks for your response! I agree that we will likely need a new RPC method. I will have a look at my current fix next week and see if I can improve it. That is: the new rpc method and implicitly call stop listener when the target is destroyed.

**Comment by jimharris** [comment]
Bug scrub: @jkryl - please post a link to your patch here once it's ready.  We still have time to get it in for the 20.01 release (code freeze in two weeks).

**Comment by jkryl** [comment]
@jimharris : I submitted a patch to gerrithub for review.

I did not add any new rpc methods as I think that would not make sense in the current situation. I just changed the implementation of delete subsystem rpc method to preserve its original behavior. That is, `nvmf_delete_subsystem` rpc method still removes all listeners as before. `nvmf_subsystem_add_listener` and `nvmf_subsystem_remove_listener` also without change. No change in behavior. But nvmf library API  follows the new approach. That is: the listener is stopped only when explicitly requested (`spdk_nvmf_tgt_stop_listen()`) or implicitly when the target is destroyed.
